File: python/chat/plugin/chat_plugin.py
from hansken_extraction_plugin.api.extraction_plugin import ExtractionPlugin
from hansken_extraction_plugin.api.extraction_trace import Tracelet
from hansken_extraction_plugin.api.plugin_info import Author, MaturityLevel, PluginId, PluginInfo

from logbook import Logger


log = Logger(__name__)


class ChatPlugin(ExtractionPlugin):

    def plugin_info(self):
        plugin_info = PluginInfo(
            id=PluginId('nfi.nl', 'chat', 'ChatPluginPython'),
            version='1.0.0',
            description='Example Extraction Plugin: Exclusive Chat format file parser',
            author=Author('The Externals', 'tester@holmes.nl', 'NFI'),
            maturity=MaturityLevel.PROOF_OF_CONCEPT,
            webpage_url='https://hansken.org',
            matcher='file.extension=txt',
            license='Apache License 2.0'
        )
        log.info('pluginInfo request')
        log.debug(f'returning plugin info: {plugin_info}')
        return plugin_info

    def process(self, trace, data_context):
        # get the name of the file
        file_name = trace.get('file.name')
        # log something to output as an example
        log.info(f"processing trace {trace.get('name')}")
        # set the chat application property on the trace
        trace.update('chatConversation.application', f'DemoApp {file_name}')

        data = trace.open().read(data_context.data_size)
        chat_messages = data.decode('utf-8').split('\n')
        log.debug(f'found chat messages: {chat_messages}')

        # each message has the format 'sender:receiver message'
        for index, chat_message in enumerate(chat_messages):
            # split contacts and message
            contacts, message = chat_message.split(' ', 1)
            # split sender and receiver
            sender, receiver = contacts.split(':')
            conversationId = [sender, receiver]
            conversationId.sort()
            conversationId = "-".join(conversationId)

            # add chat message
            child_builder = trace.child_builder(f'message {index}')

            # add a collection (tracelet of type FVT, see tracemodel for typing information)
            child_builder.add_tracelet("collection",
                    { "name": conversationId,
                      "type": "chatConversation" })

            # add two entities (tracelet of type MVT, see tracemodel for typing information)
            # (!) works with Hansken 45.19.0 or higher
            child_builder.add_tracelet("entity",
                    { "confidence": 0.76,
                      "type": "name",
                      "value": sender})
            child_builder.add_tracelet("entity",
                    { "confidence": 0.79,
                      "type": "name",
                      "value": receiver})

            child_builder.update({
                'chatMessage.application': 'DemoApp',
                'chatMessage.from': sender,
                'chatMessage.to': [receiver],  # list, because there can be multiple receivers
                'chatMessage.message': message,
            }).build()

            # add contacts as children of each message (they are the same for each message in the log,
            # but it just shows an example)
            child_builder.child_builder(sender).update({
                'contact.application': 'DemoApp',
                'contact.name': sender
            }).build()

            child_builder.child_builder(receiver).update({
                'contact.application': 'DemoApp',
                'contact.name': receiver
            }).build()

            log.debug(f'found chat message ({sender}->{receiver}): {message}')

File: java/chat/src/main/java/org/hansken/plugin/extraction/ChatPlugin.java
package org.hansken.plugin.extraction;

import static java.lang.String.format;
import static java.nio.charset.StandardCharsets.UTF_8;
import static java.util.Collections.singletonList;

import java.io.IOException;

import org.hansken.plugin.extraction.api.Author;
import org.hansken.plugin.extraction.api.DataContext;
import org.hansken.plugin.extraction.api.ExtractionPlugin;
import org.hansken.plugin.extraction.api.MaturityLevel;
import org.hansken.plugin.extraction.api.PluginId;
import org.hansken.plugin.extraction.api.PluginInfo;
import org.hansken.plugin.extraction.api.RandomAccessData;
import org.hansken.plugin.extraction.api.Trace;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class ChatPlugin implements ExtractionPlugin {

    private static final Logger LOG = LoggerFactory.getLogger(ChatPlugin.class);

    private static final String TOOL_DOMAIN = "nfi.nl";
    private static final String TOOL_CATEGORY = "chat";
    private static final String TOOL_NAME = "ChatPluginJava";
    private static final String TOOL_LICENSE = "Apache License, Version 2.0";

    @Override
    public PluginInfo pluginInfo() {
        final Author author = Author.builder()
            .name("The Externals")
            .email("tester@holmes.nl")
            .organisation("NFI")
            .build();

        return PluginInfo.builderFor(this)
            .pluginVersion("1.0.0")
            .description("Example Extraction Plugin: Exclusive Chat format file parser")
            .author(author)
            .maturityLevel(MaturityLevel.PROOF_OF_CONCEPT)
            .webpageUrl("https://hansken.org")
            .hqlMatcher("file.extension=txt")
            .id(new PluginId(TOOL_DOMAIN, TOOL_CATEGORY, TOOL_NAME))
            .license(TOOL_LICENSE)
            .build();
    }

    @Override
    public void process(final Trace trace, final DataContext dataContext) throws IOException {
        // get the name of the file
        final String fileName = trace.get("file.name");
        // log something to the output as an example
        LOG.error("processing trace " + trace.get("name"));
        // set the chat application property on the trace
        trace.addType("chatConversation").set("chatConversation.application", format("DemoApp %s", fileName));

        final RandomAccessData data = dataContext.data();
        final String[] chatMessages = new String(data.readNBytes((int) data.remaining()), UTF_8).split("\n");

        // each message has the format 'sender:receiver message'
        for (int index = 0; index < chatMessages.length; index++) {
            // split contacts and message
            final String[] contactsAndMessage = chatMessages[index].split(" ", 2);
            // split sender and receiver
            final String[] senderAndReceiver = contactsAndMessage[0].split(":");

            final String sender = senderAndReceiver[0];
            final String receiver = senderAndReceiver[1];
            final String message = contactsAndMessage[1];
            final String conversationId = sender.compareTo(receiver) < 0 ?
                sender + "-" + receiver :
                receiver + "-" + sender;

            // add chat message
            trace.newChild(format("message %d", index), messageTrace -> {
                messageTrace.addType("chatMessage")
                    .set("chatMessage.application", "DemoApp")
                    .set("chatMessage.from", sender)
                    .set("chatMessage.to", singletonList(receiver))  // list, because there can be multiple receivers
                    .set("chatMessage.message", message);

                // add a collection (tracelet of type FVT, see tracemodel for typing information)
                messageTrace.addTracelet("collection", tracelet -> tracelet
                    .set("name", conversationId)
                    .set("type", "chatConversation"));

                // add two entities (tracelet of type MVT, see tracemodel for typing information)
                // (!) works with Hansken 45.19.0 or higher
                messageTrace.addTracelet("entity", tracelet -> tracelet
                    .set("confidence", 0.76)
                    .set("type", "name")
                    .set("value", sender));

                messageTrace.addTracelet("entity", tracelet -> tracelet
                    .set("confidence", 0.79)
                    .set("type", "name")
                    .set("value", receiver));


                // add contacts as children of each message (they are the same for each message in the log,
                // but it just shows an example)
                messageTrace.newChild(sender, contactTrace -> {
                    contactTrace.addType("contact")
                        .set("contact.application", "DemoApp")
                        .set("contact.name", sender);
                });
                messageTrace.newChild(receiver, contactTrace -> {
                    contactTrace.addType("contact")
                        .set("contact.application", "DemoApp")
                        .set("contact.name", receiver);
                });
            });
        }
    }
}
